### PR TITLE
store/tikv: make snapshot thread safe (#17593)

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Cherry pick https://github.com/pingcap/tidb/pull/17593 to release-4.0

Problem Summary:

A correct fix need to cherry pick a series of commits:

#16056 
#17555 
#17593 
#17678 
#17911 

That maybe too much for 4.0

### What is changed and how it works?

make snapshot thread safe

### Check List 

Tests 

- Unit test

### Release note 

-  No release note

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @coocood @lysu

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
@tiancaiamao 
Just move code from Next to Open doesn't completely fix the bug.
The PointGet executor maybe built many times during execution. for example, in IndexJoin.

**Comment by tiancaiamao** [comment]
> @tiancaiamao
> Just move code from Next to Open doesn't completely fix the bug.
> The PointGet executor maybe built many times during execution. for example, in IndexJoin.

Move code from `Next` to `Open` so the snapshot is not captured everytime and reduce conflict (the transaction is modified and snapshot may be nil)

This code is not safe, as txn maybe set to `nil` after `Valid()` and before `GetSnapshot`.

```
if txn.Valid() {
     snapshot = txn.GetSnapshot()
}
```

I override GetSnapshot to provide the snapshot.
Not in a atomic way, but it should work , as a pointer is in a machine word size (assumption on hardware), either nil or a kv.Transaction is OK.

**Comment by tiancaiamao** [comment]
github is unavailable

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
@bobotu PTAL

File: session/txn.go [hunk 1]
```diff
@@ -309,10 +318,17 @@ func (st *TxnState) Get(ctx context.Context, k kv.Key) ([]byte, error) {
 
 // GetMemBuffer overrides the Transaction interface.
 func (st *TxnState) GetMemBuffer() kv.MemBuffer {
+	txn := st.Transaction
```
Comment (coocood) [inline session/txn.go:321]: If we check InTxn() is true, then we don't need to modify this function.

File: session/txn.go [hunk 2]
```diff
@@ -310,7 +319,7 @@ func (st *TxnState) Get(ctx context.Context, k kv.Key) ([]byte, error) {
 // GetMemBuffer overrides the Transaction interface.
 func (st *TxnState) GetMemBuffer() kv.MemBuffer {
 	if st.stmtBuf == nil || st.stmtBuf.Size() == 0 {
-		return st.Transaction.GetMemBuffer()
+		return txn.GetMemBuffer()
```
Comment (coocood) [inline session/txn.go:322]: revert this line

**Comment by coocood** [comment]
LGTM

**Pull request merged**